src/Firebase/Messaging/RequestFactory.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Kreait\Firebase\Messaging;
+
+use Beste\Json;
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+
+/**
+ * @internal
+ */
+final class RequestFactory
+{
+    private bool $environmentSupportsHTTP2;
+
+    public function __construct(
+        private readonly RequestFactoryInterface $requestFactory,
+        private readonly StreamFactoryInterface $streamFactory,
+    ) {
+        $this->environmentSupportsHTTP2 = self::environmentSupportsHTTP2();
+    }
+
+    public function createRequest(Message $message, string $projectId, bool $validateOnly): RequestInterface
+    {
+        $request = $this->requestFactory
+            ->createRequest(
+                'POST',
+                'https://fcm.googleapis.com/v1/projects/'.$projectId.'/messages:send',
+            )
+        ;
+
+        if ($this->environmentSupportsHTTP2) {
+            $request = $request->withProtocolVersion('2.0');
+        }
+
+        $payload = ['message' => $message];
+
+        if ($validateOnly === true) {
+            $payload['validate_only'] = true;
+        }
+
+        $body = $this->streamFactory->createStream(Json::encode($payload));
+
+        return $request
+            ->withBody($body)
+            ->withHeader('Content-Type', 'application/json; charset=UTF-8')
+            ->withHeader('Content-Length', (string) $body->getSize())
+        ;
+    }
+
+    /**
+     * @see https://github.com/microsoftgraph/msgraph-sdk-php/issues/854
+     * @see https://github.com/microsoftgraph/msgraph-sdk-php/pull/1120
+     */
+    private static function environmentSupportsHTTP2(): bool
+    {
+        $features = curl_version()["features"] ?? null;
+
+        return
+            extension_loaded('curl')
```

Comment:
Shouldn't we check for the loaded extension before calling `curl_version()`?

Reply:
Ah, you're absolutely right! Will fix!

Reply:
Fixed!

BTW, the reason why the method is more verbose than the one in https://github.com/microsoftgraph/msgraph-sdk-php/pull/1120 is so that PHPStan doesn't complain about `curl_version()["features"]` returning `array|false`.